core/src/main/java/org/apache/spark/shuffle/checksum/RowBasedChecksum.scala

```diff
@@ -0,0 +1,114 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.checksum
+
+import java.io.ObjectOutputStream
+import java.util.zip.Checksum
+
+import scala.util.control.NonFatal
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.network.shuffle.checksum.ShuffleChecksumHelper
+import org.apache.spark.util.MyByteArrayOutputStream
+
+/**
+ * A class for computing checksum for input (key, value) pairs. The checksum is independent of
+ * the order of the input (key, value) pairs. It is done by computing a checksum for each row
+ * first, and then computing the XOR for all the row checksums.
+ */
+abstract class RowBasedChecksum() extends Serializable with Logging {
+  private var hasError: Boolean = false
+  private var checksumValue: Long = 0
+  /** Returns the checksum value computed. Tt returns the default checksum value (0) if there
+   * are any errors encountered during the checksum computation.
+   */
+  def getValue: Long = {
+    if (!hasError) checksumValue else 0
+  }
+
+  /** Updates the row-based checksum with the given (key, value) pair */
+  def update(key: Any, value: Any): Unit = {
+    if (!hasError) {
+      try {
+        val rowChecksumValue = calculateRowChecksum(key, value)
+        checksumValue = checksumValue ^ rowChecksumValue
```

Comment:
XOR has problems when the same (key, value) pair are used multiple times. Should we track the number of pairs as well?

Reply:
I think this is a good point but hard to compute as this will be a bit more stateful. 

An interesting corner case when we have multiples of two different key value pairs. Then we cannot detect differences in those if it is different in the repetition, like
```
N times of ("key1", "value1")
M times of ("key2", "value2")
```

cannot be distinguished from for every N>=1 and M>=1  from:
```
K times of ("key1", "value1")
L times of ("key2", "value2")
``` 
for every K>=1 and L>=1 (meanwhile N+M can be even equal with K+L).

Is not it?

What about in addition to the bitwise OR (currently `checksumValue `) calculating a SUM as well and when the `getValue` is called combine those two into one number with an extra OR?

Reply:
What about in addition to the bitwise OR (currently `checksumValue `) calculating a SUM as well and when the `getValue` is called combine those two into one number with an extra OR?

Reply:
Yeah, that's what I was thinking of. But combining the number of pairs (count, not the sum) into the final checksum should be fine.

Reply:
Perhaps something like this might work ? https://en.wikipedia.org/wiki/Fowler%E2%80%93Noll%E2%80%93Vo_hash_function

It will be more expensive than xor, but should handle order and duplication.

Reply:
You are right, order sensitivity matter at beginning of task, not at end !

Sum + xor or sum + xor + multiplication with some xor folding to generate final hash might be cheap.
Can't think of other alternatives which might work well and yet is reasonably robust to duplication

Reply:
Hi folks, I am working with Jiexing to follow up on the PR.

Do you think something like below combining both the `sum` and `xor` of the hashcode would be helpful to address the concerns? cc @peter-toth @attilapiros @mridulm @cloud-fan

```
private var checksumValue: Long = 0
private var sum: Long = 0

def rotateLeft(value: Long, k: Int): Long = {
  ((value << k) & 0xffffffffffffffff) | (x >>> (64 - k))
}

def getValue: Long = {
  if (!hasError) {
    checksumValue ^ rotateLeft(sum, 27)
  } else {
    0
  }
}

def update(key: Any, value: Any): Unit = {
  ...
  val rowChecksumValue = calculateRowChecksum(key, value)
  checksumValue = checksumValue ^ rowChecksumValue
  sum += rowChecksumValue
  ...
}
```